Support optional dependencies

AlexMSmithCA opened the pull request:
Fixes #95.  Implements proposed **Option 1**:

```js
// Defining the token
import {createToken} from 'fusion-tokens';
import type {Token} from 'fusion-tokens';

type Fetch = (input: string | Request, init?: RequestOptions) => Promise<Response>;
export const FetchToken: Token<Fetch> = createTaoken('FetchToken');

// Using the token
import {createPlugin} from 'fusion-core';
export default createPlugin({
  deps: {
    fetch: FetchToken, // required token example
    logger: LoggerToken.optional, // optional token example
  },
});
```

AlexMSmithCA commented on src/create-token.js:
```diff
@@ -0,0 +1,32 @@
+// @flow
```
This currently lives in [fusion-tokens](https://github.com/fusionjs/fusion-tokens/blob/master/src/index.js), but may make more sense living here.  Alternatively, we can let `fusion-core` depend on `fusion-tokens` if we want to minimize breaking changes.

KevinGrandon replied:
Doesn't matter to me, but just don't want to duplicate code :)

If we can do this in a way that doesn't break current non-optional token usage that would be ideal I think.

lhorie replied:
I think it should live here and have fusion-tokens be only common tokens

AlexMSmithCA commented on src/base-app.js:
```diff
@@ -32,59 +33,65 @@ class FusionApp {
-      let {value, aliases, enhancers} = registered.get(token) || {};
-      // the type was never registered, throw error
+
+      // Base: the type was never registered, throw error or provide undefined if optional
```
Lines 69 through 84 deal with resolving the new optional tokens.

AlexMSmithCA commented on src/base-app.js:
```diff
@@ -32,59 +33,65 @@ class FusionApp {
+    const registered = this.registered; // Token.ref || Token -> {value, aliases, enhancers}
+    const resolvedPlugins = []; // Plugins
+    const allAliases = new Set(); // Token.ref || Token
     const resolveToken = (token, tokenAliases) => {
```
Unfortunately `token` here can be either a `Token` (new type) or any arbitrary value (which likely includes `Plugins`).  As such, we have to check truthy-ness of `token.ref` and `token` which isn't particularly elegant.  Open to suggestions on how to clean this up since this code is fairly complex already.

lhorie replied:
I'd argue for having `createToken` in this package, and doing an `instanceof` check

AlexMSmithCA replied:
+1, we could do something like:

```js
// Helpers
function getTokenRef(token: TokenImpl | any): any {
   if(token instanceof TokenImpl) {
      return token.ref;
   }
   return token;
}

And replace each `token.ref || token` with `getTokenRef(token)`. 
```

AlexMSmithCA commented on flow.js:
```diff
@@ -36,18 +42,27 @@ declare class FusionApp {
-  register<Token: number>(token: Token, val: number): aliaser<*>;
-  register<Token: Object>(token: Token, val: $Exact<Token>): aliaser<*>;
-  middleware<Deps>(deps: Deps, middleware: (Deps) => Middleware): void;
+  // register<Token: null>(token: Token, val: null): aliaser<*>;
```
Thoughts on no longer supporting arbitrary tokens (i.e. those not created via `createToken`)?  While the freedom is nice, this adds a layer of complexity (more so with optional dependencies) that I'm not convinced is worthwhile.

ganemone replied:
+1

lhorie replied:
One could make an argument for string tokens, but definitely nothing else

AlexMSmithCA commented on src/__tests__/dependency-resolution.js:
```diff
@@ -488,43 +462,43 @@ tape('dependency registration with circular dependency', t => {
-  app.register(ObjectToken, {a: 'some-d'});
-  app.resolve();
-});
+// tape('dependency configuration', t => {
```
See [comment above](https://github.com/fusionjs/fusion-core/pull/124/files#diff-877e28e5e548e988a1bc394588eccc56R55) for why this is not currently supported.

KevinGrandon commented:
Nice, you'll want to change the base branch to `master` though. The di branch was merged in earlier today.

AlexMSmithCA commented:
Added the following [changes](https://github.com/fusionjs/fusion-core/pull/124/commits/235f8c7b338ebf580929c1d8b4d6e3eb44132492):

* Updated Flow definitions
* Removed dead code in unit tests
* Helper for `getTokenRef`
* Export `createToken` and `Token` types
